File: sliding mode control/sliding_mode_control.py
"""
Sliding Mode Control

Problem Setup

We want the motor speed to be 50 rad/s

There’s some unknown push/pull (disturbance) on the motor

The SMC should keep the speed close to 50 despite this

"""

import numpy as np
import matplotlib.pyplot as plt

# Time setup
dt = 0.001
time = np.arange(0, 3, dt)

# Motor parameters
J = 0.01  # inertia
b = 0.1  # damping

# Desired speed
target_speed = 50.0  # rad/s

# SMC parameters
lambda_s = 20.0
K = 0.5

# Initial values
speed = 0.0
speed_change = 0.0

# Data logs
speeds = []
disturbances = []

for t in time:
    # Disturbance (changes halfway)
    disturbance = 0.2 if t < 1.5 else -0.4

    # Error and sliding surface
    error = target_speed - speed
    error_dot = -speed_change
    s = lambda_s * error + error_dot

    # Simple SMC control
    control = J * (-lambda_s * error_dot) + b * target_speed + K * np.sign(s)

    # Motor physics update
    speed_change = (control - b * speed + disturbance) / J
    speed += speed_change * dt

    # Save data
    speeds.append(speed)
    disturbances.append(disturbance)

# Plot speed result
plt.plot(time, speeds, label="Motor Speed")
plt.axhline(target_speed, color="k", linestyle="--", label="Target Speed")
plt.xlabel("Time (s)")
plt.ylabel("Speed (rad/s)")
plt.legend()
plt.grid()
plt.show()

File: sliding mode control/sliding_mode_control.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>

int main() {
    // Time setup
    double dt = 0.001;
    double T = 3.0;
    int steps = static_cast<int>(T / dt);

    // Motor parameters
    double J = 0.01;   // inertia
    double b = 0.1;    // damping

    // Desired speed
    double target_speed = 50.0; // rad/s

    // SMC parameters
    double lambda_s = 20.0;
    double K = 0.5;

    // Initial state
    double speed = 0.0;
    double speed_change = 0.0;

    // Data storage for plotting
    std::vector<double> time_log;
    std::vector<double> speed_log;
    std::vector<double> disturbance_log;

    for (int i = 0; i < steps; i++) {
        double t = i * dt;

        // Disturbance (changes halfway)
        double disturbance = (t < 1.5) ? 0.2 : -0.4;

        // Error and sliding surface
        double error = target_speed - speed;
        double error_dot = -speed_change;
        double s = lambda_s * error + error_dot;

        // SMC control law
        double control = J * (-lambda_s * error_dot) + b * target_speed + K * ((s > 0) - (s < 0));

        // Motor dynamics update
        speed_change = (control - b * speed + disturbance) / J;
        speed += speed_change * dt;

        // Save data
        time_log.push_back(t);
        speed_log.push_back(speed);
        disturbance_log.push_back(disturbance);
    }

    // Save to CSV for plotting in Python/Excel
    std::ofstream file("smc_motor_data.csv");
    file << "time,speed,disturbance\n";
    for (size_t i = 0; i < time_log.size(); i++) {
        file << time_log[i] << "," << speed_log[i] << "," << disturbance_log[i] << "\n";
    }
    file.close();

    std::cout << "Simulation complete. Data saved to smc_motor_data.csv\n";
    std::cout << "You can plot this in Python or Excel.\n";

    return 0;
}
